src/libostree/ostree-repo-commit.c

```diff
@@ -589,6 +589,132 @@ create_regular_tmpfile_linkable_with_content (OstreeRepo *self,
+                      GCancellable *cancellable,
+                      GError       **error)
+{
+  g_autofree char *payload_checksum = NULL;
+  g_autoptr(GInputStream) is = NULL;
+  glnx_unref_object OtChecksumInstream *checksum_payload = NULL;
+  g_autoptr(GFileInfo) file_info = NULL;
+
+  if (!G_IN_SET(self->mode, OSTREE_REPO_MODE_BARE, OSTREE_REPO_MODE_BARE_USER, OSTREE_REPO_MODE_BARE_USER_ONLY))
+    return TRUE;
+
+  if (!ostree_repo_load_file (source, checksum, &is, &file_info, NULL, cancellable, error))
+    return FALSE;
+
+  if (g_file_info_get_file_type (file_info) != G_FILE_TYPE_REGULAR
+      || g_file_info_get_size (file_info) < self->payload_link_threshold)
+    return TRUE;
+
+  checksum_payload = ot_checksum_instream_new (is, G_CHECKSUM_SHA256);
+
+  guint64 remaining = g_file_info_get_size (file_info);
+  while (remaining)
+    {
+      char buf[8192];
+      gssize ret = g_input_stream_read ((GInputStream *) checksum_payload, buf,
+                                        MIN (sizeof (buf), remaining), cancellable, error);
+      if (ret < 0)
+        return FALSE;
+      remaining -= ret;
+    }
+  payload_checksum = ot_checksum_instream_get_string (checksum_payload);
+
+  return _create_payload_link (self, checksum, payload_checksum, file_info, cancellable, error);
+}
+
+static gboolean
+_lookup_payload_link (OstreeRepo   *self,
+                      const char   *actual_payload_checksum,
+                      GFileInfo    *file_info,
+                      GLnxTmpfile  *tmpf,
+                      GCancellable *cancellable,
+                      GError       **error)
+{
+  int dfd_searches[] = { -1, self->objects_dir_fd };
+  if (self->commit_stagedir.initialized)
+    dfd_searches[0] = self->commit_stagedir.fd;
+  for (guint i = 0; i < G_N_ELEMENTS (dfd_searches); i++)
+    {
+      glnx_autofd int fdf = -1;
+      char loose_path_buf[_OSTREE_LOOSE_PATH_MAX];
+      char loose_path_target_buf[_OSTREE_LOOSE_PATH_MAX];
+      char target_checksum[OSTREE_SHA256_STRING_LEN+1];
+      int dfd = dfd_searches[i];
+      if (dfd == -1)
+        continue;
+
+      _ostree_loose_path (loose_path_buf, actual_payload_checksum, OSTREE_OBJECT_TYPE_PAYLOAD_LINK, self->mode);
+
+      if (TEMP_FAILURE_RETRY (readlinkat (dfd, loose_path_buf, target_checksum, sizeof (target_checksum))) < 0)
+        {
+          if (errno == ENOENT)
+            continue;
+          return glnx_throw_errno (error);
+        }
+
+      _ostree_loose_path (loose_path_target_buf, target_checksum, OSTREE_OBJECT_TYPE_FILE, self->mode);
+      if (!ot_openat_ignore_enoent (dfd, loose_path_target_buf, &fdf, error))
+        return FALSE;
+
+      if (fdf < 0)
+        {
+          /* If the link is referring to an object that doesn't exist anymore in the repository, just unlink it.  */
+          if (!glnx_unlinkat (dfd, loose_path_buf, 0, error))
+            return FALSE;
+        }
+      else
+        {
+          if (ftruncate (tmpf->fd, 0) < 0)
```

Comment:
Sort of unfortunate to have done the whole write until now, then toss it but I can't think of a better way.  This could at least use a comment like `/* This undoes all of the previous writes; we want to generate reflinked data */`.

Actually also...if the filesystem doesn't have reflinks, this whole thing isn't actually useful right?  It feels to me like we need to detect this property for the repo and conditionalize use of payload links at all based on that?

Reply:
by default this option is disabled, so it must be manually enabled before we start looking at the payload